quinn-proto/src/connection/assembler.rs

```diff
@@ -99,32 +107,60 @@ impl Assembler {
         let new = BinaryHeap::with_capacity(self.data.len());
         let old = mem::replace(&mut self.data, new);
         for chunk in old.into_sorted_vec().into_iter().rev() {
+            if chunk.size >= chunk.allocation_size {
+                self.data.push(chunk);
+                continue;
+            }
             let end = offset + (buffer.len() as u64);
             if let Some(overlap) = end.checked_sub(chunk.offset) {
                 if let Some(bytes) = chunk.bytes.get(overlap as usize..) {
                     buffer.extend_from_slice(bytes);
                 }
             } else {
                 let bytes = buffer.split().freeze();
```

Comment:
With the current BytesMut issues (freeze being non free, and and resizing after split allocating too much capacity - see tokio-rs/bytes#435 ) this doesn't seem an efficient strategy. I think at this point I would recommend allocating a new BytesMut whenever we detect a new fragmented section begins, and allocating only the capacity for the next fragment(s).

Reply:
Hm, I could be wrong, but I looked into the code of `BytesMut`, and for me it looks like we should be fine, because we call `split` before `freeze` which calls `promote_to_shared` and changes the kind to `KIND_ARC`. Therefore `freeze` should be free, I think. And also `promote_to_shared` does not seem to call `into_boxed_slice` like `freeze` does.

Reply:
Ok, I think you are right regarding `freeze`. Thanks for double-checking. 
And the bad allocation behavior is partially mitigated by allocating  everything upfront. However we might be over-allocating, since the initial filter doesn't check for duplicated ranges.

Reply:
Over-allocation via duplicated ranges is now fixed by calling `deduplicate()` at the beginning of `defragment`.